source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,622 @@
+  while (ir_context->cfg()->preds(convergence_block_id).size() == 1) {
+    if (convergence_block_id == header->id()) {
+      // There is a chain of blocks with one predecessor from the header block
+      // to the merge block. This means that the region is not single-entry,
+      // single-exit (because the merge block is only reached by one of the two
+      // branches).
+      return false;
+    }
+    convergence_block_id = ir_context->cfg()->preds(convergence_block_id)[0];
+  }
+
+  // Get all the blocks reachable by the header block before reaching the
+  // convergence block and check that, for each of these blocks:
+  //  - the header dominates it and the convergence block postdominates it (so
+  //    that the header and merge block form a single-entry, single-exit
+  //    region)
+  //  - it does not contain merge instructions
+  //  - it branches unconditionally to another block
+  //  - it does not contain atomic or barrier instructions
+  auto enclosing_function = header->GetParent();
+  auto dominator_analysis =
+      ir_context->GetDominatorAnalysis(enclosing_function);
+  auto postdominator_analysis =
+      ir_context->GetPostDominatorAnalysis(enclosing_function);
+
+  // Traverse the CFG starting from the header and check all the blocks that can
+  // be reached by the header before reaching the convergence block.
+  std::list<uint32_t> to_check;
+  header->ForEachSuccessorLabel(
+      [&to_check](uint32_t label) { to_check.push_back(label); });
+
+  while (!to_check.empty()) {
+    uint32_t block_id = to_check.front();
+    to_check.pop_front();
+
+    if (block_id == convergence_block_id) {
+      // We have reached the convergence block, we don't need to consider its
+      // successors.
+      continue;
+    }
+
+    // If the block is not dominated by the header or it is not postdominated by
+    // the convergence_block, this is not a single-entry, single-exit region.
+    if (!dominator_analysis->Dominates(header->id(), block_id) ||
+        !postdominator_analysis->Dominates(convergence_block_id, block_id)) {
+      return false;
+    }
+
+    auto block = ir_context->cfg()->block(block_id);
+
+    // The block must not have a merge instruction, because inner constructs are
+    // not allowed.
+    if (block->GetMergeInst()) {
+      return false;
+    }
+
+    // We need to make sure that OpSampledImage instructions will not be
+    // separated from their use, as they need to be in the same block.
+
+    // All result ids of an OpSampledImage instruction occurring before the last
+    // point where the block will need to be split.
+    std::set<uint32_t> sampled_image_result_ids_before_split;
+
+    // All result ids of an OpSampledImage instruction occurring after the last
+    // point where the block will need to be split. They can still be used.
+    std::set<uint32_t> sampled_image_result_ids_after_split;
+
+    // Check all of the instructions in the block.
```

Comment:
This all looks really good and I am impressed with your approach and the code clarity.

However, I think we might need to be even more careful. There are a lot of instructions that might cause trouble if they execute unconditionally because they have side-effects or undefined behavior. E.g. 

OpImageFetch
OpImageRead
OpImageWrite
OpArrayLength

OpAccessChain (does a non-constant index need to be in bounds? I think probably not, despite what the spec says (the spec will be updated eventually I think), so this instruction might be OK)


These are a bit like OpLoad and OpStore:
OpCopyMemory
OpCopyMemorySized

There are some instructions from the [GLSL extended instructions](https://www.khronos.org/registry/spir-v/specs/1.0/GLSL.std.450.html) that write their result to a pointer (but most don't):

Modf
Frexp

InterpolateAtSample (requires a capability)
...

And there might be other instructions that we have missed or that get added.

One approach is to explicitly list all instructions that we want to handle. I believe @Vasniktel did this for his PropagateInstructionUp transformation. See TransformationPropagateInstructionUp::IsOpcodeSupported. I just realised that this misses a lot of GLSL extended instructions (opcode is always OpExtInst, but then there is a second opcode for the GLSL instruction), so we should be sure to include those if we use this approach. 

Another approach is to conservatively assume that all instructions that have pointer or image or sampler in-operands are tricky. Also, all barrier instructions. We can then explicitly list the instructions that have pointer in-operands that are know are fine (e.g. OpAccessChain...I think...there might be a better example). Most of the tricky instructions can still be handled though; we just need to add a conditional branch like you do for OpLoad, OpStore, OpFunctionCall. For example, I think we could handle atomic instructions (and there would be no special code for this; it would be the same code that handles OpLoad, OpStore, ...[anything with a pointer in-operand]).

Reply:
I have now changed it so that:
- a few instructions are listed as having no side effects. They can always be executed
- barrier and OpSampledImage instructions are forbidden
- other instructions are allowed but require fresh ids (5 if they have a result id, 2 otherwise)